x/programs/rust/README.md

```diff
@@ -1,30 +1,134 @@
+- The first is quite minimal. When storing a value in the host, the
+SDK prepends a single byte representing the type of `ProgramValue`
+being stored. This single byte is necessary to inform Rust about the variable
+type when retrieving from the `host`.
+- The second area is a bit more complex and happens during a call to invoke
+another program. In this case we pass a byte array which contains the parameters
+for the external function call. To construct this we marshal all the params with
+their metadata into one final byte array. Each parameter is added in this order
   - length of the parameter in bytes(stored as a i64)
   - boolean, [1] if the parameter is an Int, [0] otherwise
   - the actual bytes of the parameter.
 
 On the Go side, we unmarshal in the same order.
 
+## Storage
+
+A Program has 2 storage types.
+
+- `Memory` in WebAssembly is a linear buffer of unsigned bytes that can be read and
+written to by both the `Program` or host.
+
+- `State` is a host module which is imported into the `Program` instance. It
+exposes access to a database store. Unlike `Memory` state can be durable.
+
+### Memory
+
+In this example we share a byte slice between our host and a `Program`. To
+achieve this, we can write our byte slice to the shared memory from the host using
+an exported functions called `alloc` and `memory`. To do this, we would first
+allocate a chunk of continuous memory of the length desired.  `Alloc` will
+return the `offset` of that slice.
+
+```go
+runtime := // create new runtime and initialize
+memory := runtime.Memory()
+
+bytes := []byte("hello world")
+bytesLen := uint64(len(bytes))
+
+// create a memory allocation 
+offset, _ := memory.Alloc(bytesLen)
+
+// copy bytes into memory allocation
+_ = memory.Write(offset, bytes)
+
+// pass the offset and length as params to a function
+// exported by the Program
+resp, _ := runtime.Call("some_function", offset, bytesLen )
+```
+
+`hello world` converted to a byte slice will represent our Program's memory for
+the example.
+```sh
+[offset=6              ][length=3   ]
+[104 101 108 108 111 32 119 111 114 108 100]
+[119 111 114] // wor
+```
+The below shows an exported `Program` function `some_function` which accepts a
+message of arbitrary length. Unlike fixed size length objects such as `Address`
+the length can be variable. So the `length` of the memory along with the
+`offset` are both required.
+
+```rust
+#[public]
+pub fn some_function(program: Program, message: i64, message_length: i64) -> i64 {
+    // read message bytes from memory
+    let mem = Memory::new(Pointer::new(message));
+    let msg = unsafe { mem.range(12) };
+    // msg = [119 111 114]
+}
+```
+          
+Each `Memory` is tightly bound to the lifecycle of the module. For this reason,
+`Memory` is non durable and can not be relied upon.
```

Comment:
Yeah I will add a note that a memory is bound to the runtime execution so it can not be relied upon outside of that. Which yes to your point could be a single exported function call. But it could be longer in program to program for example.